docs/rfcs/2025-07-07-node-deletion-api-improvement.md

```diff
@@ -0,0 +1,149 @@
+for handling concurrent operations: what happens when a drain/fill request arrives while deletion
+is in progress, and what happens when a delete request arrives while drain/fill is in progress.
+- **Persistent**: If the storage controller restarts during this long-running operation, we must
+preserve progress and automatically resume the deletion process after the storage controller
+restarts.
+- **Migrated correctly**: We cannot simply use the existing drain mechanism for nodes scheduled for
+deletion, as this would move shards to irrelevant locations. This could result in unnecessary load
+on the storage controller and inefficient resource utilization.
+- **Context-aware**: We need to improve our scheduling context awareness mechanism beyond the basic
+API to correctly calculate preferred node when using non-locked state operations.
+- **Force option**: Administrators need the ability to force immediate, non-graceful deletion when
+time constraints or emergency situations require it, bypassing the normal graceful migration
+process.
+
+The proposed solution addresses these challenges through the following components:
+
+- **New NodeLifecycle**: Introduce `NodeLifecycle::ScheduledForDeletion` to track nodes pending
+deletion. This state persists across storage controller restarts and prevents re-registration.
+- **New NodeSchedulingPolicy**: Introduce `NodeSchedulingPolicy::Deleting` to track the current
+deletion process if any.
+- **OperationTracker**: A service state object that manages all long-running operations with the
+following capabilities:
+  - Execute operations (drain/delete/fill) with proper concurrency control
+  - Cancel ongoing operation
+  - Queue deletion intents when there are another running operations
+  - Enforce operation rules (e.g., only one drain/fill operation at a time)
+  - Maintain persistent intent tracking for queued deletions
+  - Provide better thread safety than the existing `Option<OperationHandler>` approach
+- **Fast cancellation**: Implement fast cancellation for delete operations to prevent blocking
+deployment operations.
+
+### Reliability, failure modes and corner cases
+
+In case of a storage controller failure and following restart, the system behavior depends on the
+`NodeLifecycle` state:
+
+- If `NodeLifecycle` is `Active`: No action is taken for this node.
+- If `NodeLifecycle` is `Deleted`: The node will not be re-added.
+- If `NodeLifecycle` is `ScheduledForDeletion`: A deletion background task will be launched for
+this node. Since only one ongoing deletion task per storage controller is allowed, if there are
+multiple nodes with `ScheduledForDeletion` state, additional deletion tasks will be queued.
+
+In case of a pageserver node failure during deletion, the behavior depends on the `force` flag:
+- If `force` is set: The node deletion will proceed regardless of the node's availability.
+- If `force` is not set: The deletion will be cancelled after some number of retry attempts.
```

Comment:
Not sure how this is useful. If we are doing a graceful deletion, we expect the node to be alive.
I think we should have a force flag, but that should be the equivalent of the old deletion procedure:
kick off migrations and remove the node.